### ISSUE TITLE
rclone mount: race condition on subsequent file create and remove

### ISSUE THREAD

**Issue by rafal-krypa** [ncw/rclone#1181]
There is a race condition in "rclone mount" implementation regarding creation and deletion of files. If an empty file is created in mounted directory and then quickly removed, the attempt to remove the file fails:

    $ rclone mount -v -v -v  some_remote: mnt/
    # on the other terminal:
    $ touch mnt/test; rm mnt/test
    rm: cannot remove 'mnt/test': No such file or directory

The following logs are produced by rclone:

    2017/02/27 11:02:34 DEBUG : test: Dir.Lookup
    2017/02/27 11:02:34 DEBUG : test: Dir.Create
    2017/02/27 11:02:34 DEBUG : test: Dir.Create OK
    2017/02/27 11:02:34 DEBUG : File.Attr valid=1m0s ino=0 size=0 mode=-rw-r--r--
    2017/02/27 11:02:34 DEBUG : test: WriteFileHandle.Flush
    2017/02/27 11:02:34 DEBUG : test: WriteFileHandle.Flush ignoring flush on unwritten handle
    2017/02/27 11:02:34 DEBUG : File.Attr valid=1m0s ino=0 size=0 mode=-rw-r--r--
    2017/02/27 11:02:34 DEBUG : test: WriteFileHandle.Flush
    2017/02/27 11:02:34 DEBUG : test: WriteFileHandle.Flush ignoring flush on unwritten handle
    2017/02/27 11:02:34 DEBUG : test: WriteFileHandle.Release closing
    2017/02/27 11:02:34 DEBUG : test: Dir.Remove
    2017/02/27 11:02:34 ERROR : test: Dir.Remove error: no such file or directory
    2017/02/27 11:02:37 DEBUG : test: WriteFileHandle.Release OK

When given enough time between file creation and removal, rclone works fine:

    $ touch mnt/test; sleep 3; rm mnt/test

Correct logs look like this:

    2017/02/27 11:02:59 DEBUG : test: Dir.Lookup
    2017/02/27 11:02:59 DEBUG : test: Dir.Create
    2017/02/27 11:02:59 DEBUG : test: Dir.Create OK
    2017/02/27 11:02:59 DEBUG : File.Attr valid=1m0s ino=0 size=0 mode=-rw-r--r--
    2017/02/27 11:02:59 DEBUG : test: WriteFileHandle.Flush
    2017/02/27 11:02:59 DEBUG : test: WriteFileHandle.Flush ignoring flush on unwritten handle
    2017/02/27 11:02:59 DEBUG : File.Attr valid=1m0s ino=0 size=0 mode=-rw-r--r--
    2017/02/27 11:02:59 DEBUG : test: WriteFileHandle.Flush
    2017/02/27 11:02:59 DEBUG : test: WriteFileHandle.Flush ignoring flush on unwritten handle
    2017/02/27 11:02:59 DEBUG : test: WriteFileHandle.Release closing
    2017/02/27 11:03:01 DEBUG : test: WriteFileHandle.Release OK
    2017/02/27 11:03:02 DEBUG : test: Dir.Remove
    2017/02/27 11:03:02 DEBUG : test: Dir.Remove OK

This issue prevents usage of rclone mounts together with Calibre.
While trying to open or create Calibre library on an rclone mount point, Calibre tries to determine whether the directory respects case sensitivity in file names. It tests that by creating `calibre_test_case_sensitivity.txt` file, checking whether it was created, is accessible and whether `calibre_TesT_CaSe_sensitiVitY.Txt` is accessible too. Then Calibre tries to unlink the test file and proceed, but in this case unlink fails because of race condition in rclone, making Calibre unhappy.

I tried running `rclone mount` with additional options, like `--write-back-cache` or `--min-age 0`, but it didn't help.

<hr/>
<hr/>

> What is your rclone version (eg output from `rclone -V`)

Confirmed on latest master branch (6bad0ad9c474749b1d8588a1ecdb0129e47607c3).

> Which OS you are using and how many bits (eg Windows 7, 64 bit)

Linux 64 bit.

> Which cloud storage system are you using? (eg Google Drive)

Reproduced with Amacon Drive, Google Drive and Dropbox.

> The command you were trying to run (eg `rclone copy /tmp remote:tmp`)

Described above.

> A log from the command with the `-vv` flag (eg output from `rclone -vv copy /tmp remote:tmp`)

Described above.

**Comment by rafal-krypa** [comment]
I have tested the scenario you suggested, although it wouldn't fix my use case anyway. It produced some very interesting problems, different from what I reported originally. This test was performed on Amazon Drive.

    $ echo "hello" > mnt/test ; rm mnt/test
    rm: cannot remove 'mnt/test': No such file or directory
    $ ls mnt/test
    512 mnt/test
    $ cat mnt/test
    cat: mnt/test: Input/output error
    $ ls mnt/
    ### Listing of the directory, NOT including the "test" file

Here are the logs from rclone:
```
2017/03/09 09:39:42 DEBUG : test: Dir.Lookup
2017/03/09 09:39:42 DEBUG : test: Dir.Create
2017/03/09 09:39:42 DEBUG : test: Dir.Create OK
2017/03/09 09:39:42 DEBUG : File.Attr valid=1m0s ino=0 size=0 mode=-rw-r--r--
2017/03/09 09:39:42 DEBUG : test: WriteFileHandle.Flush
2017/03/09 09:39:42 DEBUG : test: WriteFileHandle.Flush ignoring flush on unwritten handle
2017/03/09 09:39:42 DEBUG : test: WriteFileHandle.Write len=6
2017/03/09 09:39:43 DEBUG : test: WriteFileHandle.Write OK (6 bytes written)
2017/03/09 09:39:43 DEBUG : test: WriteFileHandle.Flush
2017/03/09 09:39:43 DEBUG : test: Error detected after finished upload - waiting to see if object was uploaded correctly: HTTP code 504: "504 GATEWAY_TIMEOUT": no response body ("504 GATEWAY_TIMEOUT")
2017/03/09 09:39:43 DEBUG : test: Giving up waiting for object - returning original error: HTTP code 504: "504 GATEWAY_TIMEOUT": no response body ("504 GATEWAY_TIMEOUT")
2017/03/09 09:39:43 DEBUG : pacer: Rate limited, sleeping for 947.77941ms (1 consecutive low level retries)
2017/03/09 09:39:43 DEBUG : pacer: low level retry 1/1 (error HTTP code 504: "504 GATEWAY_TIMEOUT": no response body)
2017/03/09 09:39:43 ERROR : test: WriteFileHandle.Flush error: HTTP code 504: "504 GATEWAY_TIMEOUT": no response body
2017/03/09 09:39:43 DEBUG : test: WriteFileHandle.Release nothing to do
2017/03/09 09:39:43 DEBUG : File.Attr valid=1m0s ino=0 size=6 mode=-rw-r--r--
2017/03/09 09:39:43 DEBUG : test: Dir.Remove
2017/03/09 09:39:43 ERROR : test: Dir.Remove error: no such file or directory
```

The situation is worse than in case of empty file. In the case of empty file, first removal failed due to race condition. But after few seconds the file could be opened properly and removed without errors.
In this case when race happens for non-empty file, initial remove fails but then any subsequent attempt for opening or removing the file fails too.

**Comment by ncw** [comment]
@rafal-krypa that is Amazon being unreliable and rclone not retrying the upload (because in the general case it needs to buffer it to disk).    The unreliability I'm going to fix in #711.

**Comment by breunigs** [comment]
I looked into a synchronous `release`, but as far as I can tell the best approach was a patchset from around 2014 that introduced this feature – it didn't get merged into the kernel, so, no dice.

Here's another idea: create a 0 byte file when trying to flush an unwritten handle. Only once of course, unless a write happens – which then would behave like it does not and possibly overwrite a file of the same name. Obvious downside is the extra request even when not needed (for sure when using a pipe, possibly other cases). It would also most likely fix `rsync`, for better or worse.

**Comment by breunigs** [comment]
(PR was in #1993)
